### DISCUSSION TITLE
Getting email from Github oAuth

### DISCUSSION THREAD

**Discussion by junaid33** [nextauthjs/next-auth#991] [Help] [answered]
Hello,

We are using Nextauth with Prisma and things are running smoothly when use email to login. When we use Github oAuth, email is not returned if the users doesn't have the email shown publicly. Our application needs the email. I tried to use a custom Provider to see if we can get the email somehow:

```
    {
      id: 'github',
      name: 'GitHub',
      type: 'oauth',
      version: '2.0',
      scope: 'user',
      accessTokenUrl: 'https://github.com/login/oauth/access_token',
      authorizationUrl: 'https://github.com/login/oauth/authorize',
      profileUrl: 'https://api.github.com/user',
      profile: profile => {
        console.log({ profile });
        return {
          id: profile.id,
          name: profile.name || profile.login,
          email: profile.email,
          image: profile.avatar_url,
        };
      },
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    }
```

I can fetch https://api.github.com/user/emails under the profile function use use that to get the email. But how can I access the accessToken in that function?

**Comment by junaid33** [comment, marked as answer]
Found the answer. We can add a signin callback:

```
  callbacks: {
    signIn: async (user, account, metadata) => {
      if (account.provider !== 'github') return;

      const emailRes = await fetch('https://api.github.com/user/emails', {
        headers: {
          Authorization: `token ${account.accessToken}`,
        },
      });
      const emails = await emailRes.json();
      const primaryEmail = emails.find(e => e.primary).email;

      user.email = primaryEmail;
    },
  },
```

We could add this to the documentation under the Github provider since many users do need the user email.

**Reply by TimNZ** [reply]
Yes this should be added to the docs/guides to save a bunch of people a lot of time.

https://github.com/nextauthjs/next-auth/discussions/976

**Comment by imadatyatalah** [comment]
Any idea how to get login of the user and return it to the user object?
for example my login is `imadatyatalah`.
I want the user object to return it too.

**Reply by balazsorban44** [reply]
check the jwt and session callbacks

https://next-auth.js.org/configuration/callbacks

Also the profile provider callback has a second parameter now, containing tokens

https://next-auth.js.org/configuration/providers#using-a-custom-provider
